### DISCUSSION TITLE
Rebuttal to the points raised in this article (please see details below ) ?

### DISCUSSION THREAD

**Discussion by sktguha** [reduxjs/redux#3983] [Help]
What is the solution to the two problems mentioned in this article ? ( don't agree too much with the rest of the article though) ,

1) https://codeburst.io/the-ugly-side-of-redux-6591fde68200#fa23:~:text=The%20store%E2%80%99s%20life%20cycle%20is%20different%20from%20the%20components%E2%80%99%20life%20cycle

2) https://codeburst.io/the-ugly-side-of-redux-6591fde68200#a6cf:~:text=The%20global%20store%20is%20a%20singleton%2C%20but%20a%20component%20may%20have%20many%20instances

I have myself run into both problems personally.

**Comment by msutkowski** [comment]
Hi @sktguha! 

IMO, these aren't problems, but just poor design decisions. The example of multiple shopping carts is interesting, but the concept of adding booleans to track separate carts shows a lack of understanding of best practices. I would guess that if the author revisited this subject, they'd most likely want to rewrite the majority of this. What this article doesn't address with its local state approach in the e-commerce scenario is something as simple as showing a cart total + quantity in a navbar. As soon as you do that, you're going to introduce a different state solution, so basically, every point raised here is irrelevant.

In terms of modern react-redux, you could very easily implement this exact scenario with hooks and just giving each cart a temporary id and having a slice structure like `{ cart: { cart_key_1: { ... attributes } } }`. When the component is mounted, initialize a new a cart with the temporary key, and when it's gone, remove the cart from the store if that behavior is desired. This would solve the issues that are raised, but is still a non-realistic implementation and I wouldn't recommend doing that in most scenarios. 

At the end of the day, I'm pretty sure this article is just complaining about redux boilerplate, which is mostly solved by Redux Toolkit. Judging by the last update that is quoted below, I'd just disregard this whole thing.

> Update (11/09/2020): This post is pretty old now, and many things has changed since I wrote it. Today, Dan Abramov, the creator of Redux, already admits that Redux was a mistake, and Redux is loosing his popularity. Controllerim is still working and pretty much stable, but I am no longer maintaining it. My efforts are now focusing on suggesting a built-in state management solution for React itself, using Stores, Context, and local state improvements.

If you have specific issues that are relevant to what you're actively working on, I'd bring those up instead of the points in this article, as what is represented there are outdated complaints that were most likely given just to promote the author's own library. If you look at that quote above, they're now basically recommending re-implementing Redux, but most likely in a less performant manner. @markerikson has many articles on the topic of Redux vs Context that have been linked many times - here is [a relevant thread from yesterday](https://www.reddit.com/r/webdev/comments/kobhft/for_react_users_is_it_safe_to_say_that_usecontext/ghqccdx/?utm_source=reddit&utm_medium=web2x&context=3) that include such links.

If you're actively in need of help with Redux best practices and implementation approaches, I'd recommend reading [the latest fundamentals guide](https://redux.js.org/tutorials/fundamentals/part-1-overview), then checking out #redux on reactiflux where there are many people that can help you.

**Reply by sktguha** [reply]
Hello Matt @msutkowski ,
Thanks a lot for your comprehensive response.

> If you have specific issues that are relevant to what you're actively working on, I'd bring those up instead of the points in this article

Actually as mentioned in the original question, I have very recently faced both of these issues in my React Redux app I am working on 🙂 .
So that's why I brought it up. **Maybe the shopping cart example as mentioned in the article was not the best one, but in other components, for example** consider a simple todoDisplay component that fetches a todo by id from the server , like ``/api/get-todo-by-id?id=${id}`` and displays it, these two problems can crop up.

In this case we would need to manually clear the cache ( as the requirement is to fetch freshly the todos everytime the todoDisplay component is opened by user). Yes in `componentWillUnmount` we can put a special clearCache action, but we need to also add some kind of scope variable etc, like 'todoDisplayScope' to distinguish between similiar such components, because repeating the clearCache type and action for each component IMO get repetitive and verbose and is a hurdle for reusable reducers etc.

So actually I am planning to wrap such components in an HOC to avoid repeating the clearCache action on componentWillUnmount everytime. Also roughly something like this, a lightweight wrapper/abstraction over React-Redux connect ( I guess similar to the apollo client's `connect` and `mapQueriesToProps`)

```javascript
const mapNetworkCallsToProps = (ownProps, dispatchMyNwCallAction) => ({
    fetchTodos: dispatchMyNwCallAction({
        typesPrefix: 'FETCH_TODO_BY_ID',
        propNameToPassBack: 'todo',
        payloadParams: {
            todoId: ownProps.todoId
        }
    })
});
export default myConnect(mapNetworkCallsToProps)(TodoDisplay);
```
 
So the `TodoDisplay` component will call the `fetchTodos()` action in its `componentDidMount` method . It will get a prop into it, 
i.e `this.props.todo` which would contain the latest network call status i.e initial, success or error and something like `this.props.todo.todoText` , incase the request is a success ( of course the middleware and reducer logic would still need to be implemented, but now its a common one, required to be implemented only once and can be used multiple times for seperate components )

I will still use Redux underneath, but now for these kind of components, I need **not bother with thinking where in the Redux state I have to put them** as they really are not meant to be globally used by anyone else. Probably these kind of states will go somewhere common like `state.privateNetworkStates.todoDisplay_some_unique_id_per_instance` and `state.privateNetworkStates.postDisplay_some_unique_id_per_instance` etc . 

I am aware that solutions such as **React Query** by Tanner Lindsley exist, but I want to stick with Redux for now and not integrate two seperate libraries for the same/similiar thing. 

> In terms of modern react-redux, you could very easily implement this exact scenario with hooks and just giving each cart a temporary id

Might use uuid here as the temporary id to solve the different instance but same piece of state problems.

If you get some time, please let me know what do you think of this `mapNetworkCallsToProps` abstraction 🙂  ( maybe the name could be improved ).

**Reply by msutkowski** [reply]
I'd recommend taking a look at https://rtk-query-docs.netlify.app/ - it bolts right on to your existing redux implementation and has very similar features to react-query. 

I can't really give too much input on your solution, but I would advise against going down that path unless you have a really good reason to. There are a lot of redux-based libraries that try to do similar things, but they're all very opinionated in their implementations and usually built for specific purposes. 

In my own projects, I've moved entirely away from HOCs and connect usage. I'm biased _(a collaborator on the project)_, but I now use rtk-query for mostly every network request and it covers everything I could possibly want to do while integrating _very nicely_ with my existing redux codebases.

**Reply by sktguha** [reply]
Hello @msutkowski , Thanks for reply. indeed I did take a look at rtk query.
But some issues with rtk query, 
1) Does it handles the clearing cache on unmount automatically (which is one of my primary requirements). I know I can dispatch the clear cache on componentWillUnmount but wanted to automate that.

2) You still have to write your own `mapDispatchToProps` and `mapStateToProps`. I believe there is a `reducerPath` param in the createApi API of rtk query. I want to avoid that. i.e as mentioned, I don't want to bother at all with thinking where I want to put my network request results in the Redux state and read from it (some common state field like `state.privateNetworkStates.todoDisplay_some_unique_id_per_instance` ), for "simple fetch some data from network and display components and invalidate cache on unmount ( and freshly fetch data when it is opened/initialised again )"

3) Also at the moment I can't use hooks. I know they are recommended by the React team but for now have to stick with class components.  Hence the connect and HOC way.

**Reply by phryneas** [reply]
1. will be done by hooks,
2. in that case your `reducerPath` would be `privateNetworkStates`. Also, you don't have to ever use that again as you are provided with selectors by rtk-query
3. if you just aren't allowed to use hooks,  you'll probably have to go with a HOC - in the simplest case calling it a one-off where the HOC just calls the hook (if it's technically possible but just a "style guide thing") and passes everything down to child components, in the more complex case recreating the logic of the hooks in a HOC for once.

**Reply by sktguha** [reply]
1.
>will be done by hooks,

Not sure fully, but I still have to call the clearCache method that I would get from the useQuery or something manually in my componentWillUnmount method right ?

2.
> in that case your reducerPath would be privateNetworkStates. Also, you don't have to ever use that again as you are provided with selectors by rtk-query

That's really awesome. But I don't need any of the rtk query features actually at the moment like fetching every 5 secs etc. So don't want to add it for nothing, an extra layer of abstraction.

**Reply by phryneas** [reply]
> > will be done by hooks,
> 
> Not sure fully, but I still have to call the clearCache method that I would get from the useQuery or something manually in my componentWillUnmount method right ?

No, hooks are lifecycle-aware. They can do stuff on mount and they can do stuff on cleanup.

**Comment by phryneas** [comment]
Adding to all this: if you have state that is created when a component mounts and should be removed when a component unmounts, that is local component state. It belongs into your component.
It was never the goal of redux to manage that data.
[See the FAQ on that](https://redux.js.org/faq/organizing-state#do-i-have-to-put-all-my-state-into-redux-should-i-ever-use-reacts-setstate)

**Reply by sktguha** [reply]
Hi @phryneas , Thanks for your comment, 
But the data is fetched from a network request, for ex - for example a simple todoDisplay component that fetches a todo by id from the server , like /api/get-todo-by-id?id=${id} and displays it.

So short of making network requests in the component itself (which I don't want to do ), I have no choice but to involve Redux. 

Hence the motivation for my proposed solution: [see above](https://github.com/reduxjs/redux/discussions/3983#discussioncomment-258034-permalink:~:text=const%20mapNetworkCallsToProps%20%3D%20(ownProps%2C%20dispatchMyNwCallAction)%20%3D%3E%20(%7B
)

**Reply by phryneas** [reply]
Nothing keeps you from wrapping your component in a HOC that makes the network request and keeps it in the HOCs local state. Of course, hooks are nicer, but for some reason you don't seem to have access to those.

**Reply by sktguha** [reply]
> Nothing keeps you from wrapping your component in a HOC that makes the network request and keeps it in the HOCs local state.

Yes definitely. Actually that's kind of what I am planning :) . But instead of the HOC making the request, the HOC will under the hood create some kind of unique id/key and utilise the Redux state and middleware only. Several reasons for that, including the fact that we have a logger middleware, then some perf measuring store enhancer etc, also the fact that it can be easy to trigger dependant/related network requests etc from the middleware and not having to maintain two different systems for same/similiar purpose

**Reply by phryneas** [reply]
> > Nothing keeps you from wrapping your component in a HOC that makes the network request and keeps it in the HOCs local state.
> 
> Yes definitely. Actually that's kind of what I am planning :) . But instead of the HOC making the request, the HOC will under the hood create some kind of unique id/key and utilise the Redux state and middleware only. Several reasons for that, including the fact that we have a logger middleware, then some perf measuring store enhancer etc, also the fact that it can be easy to trigger dependant/related network requests etc from the middleware and not having to maintain two different systems for same/similiar purpose

In that case I'd suggest the "reimplement the rtk-query hooks as HOC" option above.
See those hooks source:
https://github.com/rtk-incubator/rtk-query/blob/next/src/react-hooks/buildHooks.ts#L177

Essentially:
onMount: `const result = dispatch(endpoint.initiate)`
onUnmount: `result.unsubscribe()`

**Reply by sktguha** [reply]
ok this seems interesting, but can these hooks work independently/ with minimum dependencies. I am not sure. I don't want to import the entire rtk query library just for this..
 
I mean with extra layer of rtk query sitting now, it can become extra hard to trace down any bugs for whatever reason they might happen.

**Reply by phryneas** [reply]
With that reasoning you won't be able to add any library ;)

RTK-query is essentially redux toolkit + another 4kb of code. If you are not using redux toolkit at the moment, I'd **highly** encourage you to check that out, as that is the way we are officially recommending everyone to use redux nowadays and it will reduce your code significantly. (See https://redux.js.org/tutorials/fundamentals/part-8-modern-redux ) 

That aside, everything in RTK query is still "normal redux", so you can observe it in the devtools at will.

Of course, it has to be added: rtk-query is still in alpha, but I don't expect any big breaking changes there, maybe a renamed option or some additional features, but nothing beyond that.

**Reply by sktguha** [reply]
yes fair enough :) . Yes I have been thinking about integrating RTK, but will have to find some dedicated time for that. 

But at the moment, I just want to make network requests. I don't need any of the rtk query features actually at the moment like polling, complex cache invalidation etc. So at this moment adding rtk query won't make sense right ?

In future the wrapper HOC implementation can be changed to use RTK query under the hood. The components using it won't need any changes/ minimal changes

**Reply by phryneas** [reply]
Well, the real killer feature for starters is maybe "cache sharing". If two components require the same remote endpoint with the same parameters, with your approach you will send two network requests if you really generate an uuid per component. And if you don't want those components to share anything I really have to wonder why you would use a globally shared cache at all then.

**Reply by sktguha** [reply]
@phryneas 
> If two components require the same remote endpoint with the same parameters, with your approach you will send two network requests if you really generate an uuid per component

well very true. That is fine. So at the moment whenever any component is opened it will freshly fetch the data. The reason is that we want the latest data as much as possible from the server incase it is changed by some other user etc ( yes this maybe a perfect candidate for polling, but polling would increase server load(ok maybe 2 power n polling could be better) but this is something I have to discuss with the backend team). I know its not ideal and say if the components were opened within 5 secs of each other, the double call would actually be for nothing, but these are bit advanced use cases and at the very moment it is not required for my case atleast.

>And if you don't want those components to share anything I really have to wonder why you would use a globally shared cache at all then

Very very valid point. I am very optimistic that with this new `mapNetworkCallsToProps` abstraction we will be able get rid of a massive amount of boilerplate from our app. only very few components actually need some kind of globally synced data. So for those I think we still require Redux for that. It is best for that purpose and can't be substituted by context API etc, as I am sure you are well aware.

**Reply by phryneas** [reply]
Sure, but see the FAQ entry above: that you are using redux does not mean that everything has to go into redux. If these api responses are local state (and it very much sounds like that), don't put them into redux, but into the local state of your HOC. You can still use redux for everything else.

**Reply by sktguha** [reply]
But then again I lose benefits like logging (done by the logger middleware) , perf measurements store enhancer etc if I use the local HOC state approach.

**Reply by sktguha** [reply]
And also easy debuggability via the Redux devtools and more. And suppose the simple component network call requires some kind of adhoc Redux state from elsewhere, for ex auth token(s) etc. It can be done easily from the middleware and I would say in certain cases better to do so.
The result of the network call is still kinda a local state for the component, but making it requires some data from elsewhere. Middleware(s) can do it very nicely IMO.

**Reply by sktguha** [reply]
Also say all network calls need to wait/lock before a `/api/init` call is successfully done, again much easier to implement in the common middleware then two different places

**Reply by markerikson** [reply]
fwiw, you could have spent all this time trying out RTK / RTK-Query and seeing if it suits your needs, rather than arguing :)

**Reply by sktguha** [reply]
oh hi Mark . Nice to see you :) .
well as mentioned earlier, I did check it out and RTK , but I don't need any of the functionality like polling , complex cache invalidation at the moment (I know , not ideal but not required at this moment)

**Reply by phryneas** [reply]
RTK and RTK-query are to distinct things ;)

**Reply by sktguha** [reply]
yes yes I know :)  .I have checked out both of them. Definitely wanna try/use them in future.

**Reply by sktguha** [reply]
Anyways thanks everyone for their time in discussing this with me 🙂.

**Reply by sktguha** [reply]
Hello @markerikson , While developing my own library, I had an idea, i.e a proposed simplification of the mapDispatchToProps API,
in React-Redux connect itself.

What if instead of putting dispatch calls, the developer simply returns the action object from their mapDispatchToProps and React-Redux will call dispatch internally , i.e

**OLD**
```javascript
import { connect } from 'react-redux';

 connect(null, (dispatch, ownProps)=>({
    fetchTodo : (...args)=>dispatch(myactions.fetchTodo({ ...args, todoId: ownPropsId}))
 }))(TodoDisplay)
```

**NEW SIMPLIFIED**
```javascript
import { connect } from 'react-redux';

 connect(null, (ownProps)=>({
    fetchTodo : (...args)=>myactions.fetchTodo({ ...args, todoId: ownPropsId})
    // no need to call dispatch call from here. it will be called internally by React-Redux
    // React Redux will get the return value of the function(s)  , as fetchTodo here, and pass it on to dispatch internally
 }))(TodoDisplay)
```

I completely understand that **there is an object shorthand form already** available. But that expects a plain object and won't handle this case. I can roughly think of some benefits of this, apart from the simplification. 
But first of all, what is your initial thoughts on this ?
cc @phryneas @msutkowski

**Reply by phryneas** [reply]
The connect API is an API that is used by probably by millions of code bases and thousands of tutorials. I don't see any way of introducing any breaking change there that would not be extremely confusing.

Also, we are recommending to use the hooks as the way to use react-redux. While the connect API will no go away, it is very unlikely to see any changes in the future, as it is more of a legacy tool than something we'll actively put time in. We expect class components to go away more and more, and thus this API being used less & less (granted, this can still be used for some edge cases like what you are doing, but for that it already does everything that is needed).
The only updates I see it getting are necessary updates to keep it working with future versions of react.

**Reply by sktguha** [reply]
Oh all right understood. btw I was not suggesting making any breaking change. But find some way so that it is not breaking.

But ok I get your point.

**Reply by sktguha** [reply]
Hi @phryneas , this is the example usage of the library that we had discussed earlier here. Calling it simply `redux-simplifier` now.
will integrate RTK Query in future to it.
https://gist.github.com/sktguha/3bbfdd5b8cbdfc4dfa6dd609060edb4d
I have not covered the middleware or reducer integration here. but those are one time only

I am using it in my project at work and in process of talking to other teams also to integrate it.
Please let me know any comments/feedback/criticism 🙂 and thanks for your time again for the earlier discussions.
